### PR TITLE
feat: add getter fns for validator data

### PR DESCRIPTION
**Pull request by josemvcerqueira**
## Description 

It adds getter functions to sui_system.move, sui_system_state_inner.move and validator_set.move to read the following information:

- voting power
- stake amount
- tallying rule (reported array)
- gas price
- commission rate

## Test Plan 

How did you test the new or updated feature?

We did not add any features. We only added read functions. Therefore, no tests were added.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/sui-system/sources/sui_system.move [hunk 1]
```diff
@@ -533,6 +533,17 @@ module sui_system::sui_system {
         sui_system_state_inner::active_validator_addresses(self)
     }
 
+    /// Returns all the validators who are currently reporting `addr`
+    public fun get_reporters_of(wrapper: &mut SuiSystemState, addr: address): VecSet<address> {
```
Comment (sblackshear) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:537]:
```suggestion
    public fun get_reporters_of(wrapper: &SuiSystemState, addr: address): VecSet<address> {
```

File: crates/sui-framework/packages/sui-system/sources/sui_system.move [hunk 2]
```diff
@@ -533,6 +533,17 @@ module sui_system::sui_system {
...
+        let self = load_system_state(wrapper);
+        sui_system_state_inner::get_reporters_of(self, addr)
+    }
+
+    /// Return the currently active validator by address
+    public fun active_validator_by_address(self: &mut SuiSystemState, validator_address: address): &Validator {
```
Comment (sblackshear) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:543]:
```suggestion
    public fun active_validator_by_address(self: &SuiSystemState, validator_address: address): &Validator {
```

File: crates/sui-framework/packages/sui-system/sources/sui_system.move [hunk 3]
```diff
@@ -596,6 +607,14 @@ module sui_system::sui_system {
         inner
     }
 
+    #[test_only]
+    /// Returns the total amount staked with `validator_addr`.
+    /// Aborts if `validator_addr` is not an active validator.
+    public fun validator_stake_amount(wrapper: &mut SuiSystemState, validator_addr: address): u64 {
```
Comment (sblackshear) [inline crates/sui-framework/packages/sui-system/sources/sui_system.move:613]:
```suggestion
    public fun validator_stake_amount(wrapper: &SuiSystemState, validator_addr: address): u64 {
```

File: crates/sui-framework/packages/sui-system/sources/sui_system.move (hunk 1 above)
Comment (josemvcerqueira) [reply crates/sui-framework/packages/sui-system/sources/sui_system.move:537]: Hello @sblackshear,

The function load_system_state requires a &mut SuiSystemState to load_inner_maybe_upgrade.
I believe this applies to all suggestions.

**Comment by emmazzz** [comment]
Hi @josemvcerqueira thank you for putting up this PR!

I noticed that for `active_validator_by_address` we are returning a reference to the type `Validator`, which is a type currently only accessible from the `sui_system` package. And we would like to keep it this way so that the type can be upgraded to V2 without causing any issues for other packages. This is also the reason why we have `ValidatorWrapper`s stored in tables (there's no way to migrate objects in tables to new types since we can't iterate through them). 

So my recommendation is to add individual getters for validator data fields that you need. For example, I imagine you may want to know the gas price quote of a validator, then you could add a getter in `sui_system` returning a u64 for gas price. I understand this is more verbose than the code you currently have but this is the best way I can think of that preserves the upgradeability of `sui_system` types. Thanks!

**Comment by josemvcerqueira** [comment]
> Hi @josemvcerqueira thank you for putting up this PR!
> 
> I noticed that for `active_validator_by_address` we are returning a reference to the type `Validator`, which is a type currently only accessible from the `sui_system` package. And we would like to keep it this way so that the type can be upgraded to V2 without causing any issues for other packages. This is also the reason why we have `ValidatorWrapper`s stored in tables (there's no way to migrate objects in tables to new types since we can't iterate through them).
> 
> So my recommendation is to add individual getters for validator data fields that you need. For example, I imagine you may want to know the gas price quote of a validator, then you could add a getter in `sui_system` returning a u64 for gas price. I understand this is more verbose than the code you currently have but this is the best way I can think of that preserves the upgradeability of `sui_system` types. Thanks!

@emmazzz 

Hello Emma, thank you so much for the detailed explanation. Finally able to fully understand the pattern used in sui_system and will use it in our LST :). I added the getter functions as you suggested.

**Comment by josemvcerqueira** [comment]
Bumping the PR @emmazzz 👍

**Comment by emmazzz** [comment]
Looks good to me in general! There is a compilation error though regarding VecSet. Do you mind fixing it? @josemvcerqueira 
Then I (or someone from our team) can take over the PR to finish with needed operations like protocol upgrade and framework snapshot generation

**Comment by josemvcerqueira** [comment]
> Looks good to me in general! There is a compilation error though regarding VecSet. Do you mind fixing it? @josemvcerqueira Then I (or someone from our team) can take over the PR to finish with needed operations like protocol upgrade and framework snapshot generation

oh Wow sorry about that. I forgot to run the sui move build command. Only ran the sui move test command :(

Thanks for all the help and instruction! Learned a lot.

**Pull request closed without merging**